### ISSUE TITLE
spatial_gradient(order=2, normalized=True) scales dxy differently from dxx/dyy, so hessian_response returns a wrong determinant

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4710]
## Summary

`spatial_gradient(..., order=2, normalized=True)` scales the mixed channel `dxy` differently from `dxx`/`dyy`, because `normalize_kernel2d` divides each kernel by its own absolute sum rather than by the factor that turns it into a derivative estimate. `hessian_response` uses the normalized default, so the determinant it returns is `fxx*fyy - k*fxy**2` with `k != 1`.

| mode | kernel abs sum `dxx` / `dxy` | normalized `dxx` on `x**2` (true 2) | normalized `dxy` on `x*y` (true 1) |
|---|---|---|---|
| `sobel` | 64 / 36 | 2.0 | 1.778 (= 64/36) |
| `diff` | 4 / 4 | 0.5 | 1.0 |

## Reproduction (measured on main 5390ec29d; the numbers are the same with #4683, which only changes signs)

```python
import torch
from kornia.feature import hessian_response
c = torch.arange(9.); y, x = torch.meshgrid(c, c, indexing="ij")
for m in ["sobel", "diff"]:
    for s, true in [(x*x + y*y, 4.0), (x*y, -1.0), (0.5*x*x + 0.5*y*y + x*y, 0.0)]:
        print(m, true, hessian_response(s[None, None], grads_mode=m)[0, 0, 4, 4].item())
```

```text
sobel  4.0 ->  4.0
sobel -1.0 -> -3.1605
sobel  0.0 -> -2.1605
diff   4.0 ->  0.25
diff  -1.0 -> -1.0
diff   0.0 -> -0.9375
```

The last surface is a parabolic cylinder (`(x+y)**2 / 2`, determinant exactly 0). Both modes score it as a strong saddle, so the relative weight of blob against saddle responses in `hessian_response` / `BlobHessian` depends on `grads_mode` and is wrong in both.

## Expected

The normalized second order channels share one scale: sobel divides all three kernels by 64 (the response of the unnormalized `dxx` kernel to `x**2/2`), and diff divides `dxy` by 4 (the central-difference denominator) and leaves `dxx`/`dyy` alone. The `order=1` path is not affected: both of its kernels share their absolute sum.

A fix changes `hessian_response` output values, so it needs a `breaking` fragment and re-checked doctests/pins in `tests/feature/test_responces_local_features.py`.

Found while reviewing #4683.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by Nicholas022400701** [comment]
I can take this one if you want it, the change is ready below.

**Change.** New `normalize_kernel2d_2nd_order` in `kornia/filters/kernels.py` divides each of `(dxx, dxy, dyy)` by the magnitude of its response to `x**2 / 2`, `x * y` and `y**2 / 2`, which is 64, 64, 64 for sobel and 1, 4, 1 for diff. `spatial_gradient` calls it for `order=2` and keeps `normalize_kernel2d` for `order=1`, so the first order path and `normalized=False` are untouched. The magnitude keeps the current signs, so the patch applies on `main` before or after #4683.

**Numbers.** `hessian_response` at the centre of the 9x9 grid from the report, both modes: `x**2 + y**2` 4.0, `x * y` -1.0, `(x + y)**2 / 2` 0.0 (before: sobel 4.0 / -3.1605 / -2.1605, diff 0.25 / -1.0 / -0.9375).

**Tests.** New `test_second_order_normalized_quadratics` in `tests/filters/test_sobel.py` checks the normalized channels on five quadratics and the determinant; it fails on `main` (greatest differences 0.78 sobel, 1.5 diff) and passes with the change. The `BlobHessian` pins in `test_responces_local_features.py` and the doctest in `hessian_response` are regenerated in float64, blob centres keep their value (0.1914, 0.0547), saddle values shrink. One more test needed a change: `test_negative_detections_sort_before_the_padding` in `test_scale_space_detector.py` ran `BlobHessian - 1.0` on `torch.rand(64, 64)`, which left 1 to 3 scale-space maxima whose survival depended on the response scale and the dtype; with the new scale float32 found none. It now runs on nine Gaussian blobs of increasing size (4 to 7 detections in float32 and float64, on `main` as well). `pytest tests/filters/test_sobel.py tests/feature/test_responces_local_features.py tests/feature/test_scale_space_detector.py tests/feature/test_integrated.py --dtype float16,bfloat16,float32,float64`: 664 passed, 46 skipped. Doctests of `sobel.py`, `kernels.py`, `responses.py`: 19 passed. ruff clean.

**Changelog.** `changelog.d/+fix-second-order-normalization.breaking.md` with the old and new determinant formulas and the table above, to be renamed to the PR number.

Say the word and I open the PR against `main`.

<details>
<summary>Full diff, 7 files, +95 -22</summary>

```diff
diff --git a/changelog.d/+fix-second-order-normalization.breaking.md b/changelog.d/+fix-second-order-normalization.breaking.md
new file mode 100644
index 0000000..0879287
--- /dev/null
+++ b/changelog.d/+fix-second-order-normalization.breaking.md
@@ -0,0 +1,19 @@
+`spatial_gradient(..., order=2, normalized=True)` now scales every second order kernel by its response to the
+quadratic it estimates, so `dxx`, `dxy` and `dyy` come out in the same units: on `x**2 / 2`, `x * y` and
+`y**2 / 2` each channel returns 1 in magnitude in both modes. Previously each kernel was divided by its own
+absolute sum, which gave the mixed channel a scale of its own: `mode="sobel"` returned `dxy` too large by
+`64 / 36`, `mode="diff"` returned `dxx` and `dyy` too small by 4. `order=1` and `normalized=False` are
+unchanged.
+
+`hessian_response` and `BlobHessian` use the normalized default, so their values change: the determinant is
+now `fxx * fyy - fxy**2` instead of `fxx * fyy - (64 / 36)**2 * fxy**2` (sobel) or `fxx * fyy / 4 - fxy**2`
+(diff). At the centre of a 9x9 grid:
+
+| surface (true determinant) | sobel before | sobel after | diff before | diff after |
+|---|---|---|---|---|
+| `x**2 + y**2` (4) | 4.0 | 4.0 | 0.25 | 4.0 |
+| `x * y` (-1) | -3.1605 | -1.0 | -1.0 | -1.0 |
+| `(x + y)**2 / 2` (0) | -2.1605 | 0.0 | -0.9375 | 0.0 |
+
+Blob maxima keep their value, saddle responses shrink, so the ranking of `ScaleSpaceDetector` and
+`MultiResolutionDetector` with `BlobHessian` can change between blob and saddle candidates. Refs #4710.
diff --git a/kornia/feature/responses.py b/kornia/feature/responses.py
index 5e8ee8d..9d00d5e 100644
--- a/kornia/feature/responses.py
+++ b/kornia/feature/responses.py
@@ -248,13 +248,13 @@ def hessian_response(
         ... ]]])  # 1x1x7x7
         >>> # compute the response map; rounded for a platform-stable display
         >>> hessian_response(input).round(decimals=4) + 0.0
-        tensor([[[[-0.0564, -0.0759, -0.0253,  0.0000, -0.0253, -0.0759, -0.0564],
-                  [-0.0759, -0.0330,  0.0333,  0.0000,  0.0333, -0.0330, -0.0759],
-                  [-0.0253,  0.0333,  0.0542,  0.0000,  0.0542,  0.0333, -0.0253],
+        tensor([[[[-0.0137, -0.0332, -0.0205,  0.0000, -0.0205, -0.0332, -0.0137],
+                  [-0.0332,  0.0098,  0.0381,  0.0000,  0.0381,  0.0098, -0.0332],
+                  [-0.0205,  0.0381,  0.0547,  0.0000,  0.0547,  0.0381, -0.0205],
                   [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
-                  [-0.0253,  0.0333,  0.0542,  0.0000,  0.0542,  0.0333, -0.0253],
-                  [-0.0759, -0.0330,  0.0333,  0.0000,  0.0333, -0.0330, -0.0759],
-                  [-0.0564, -0.0759, -0.0253,  0.0000, -0.0253, -0.0759, -0.0564]]]])
+                  [-0.0205,  0.0381,  0.0547,  0.0000,  0.0547,  0.0381, -0.0205],
+                  [-0.0332,  0.0098,  0.0381,  0.0000,  0.0381,  0.0098, -0.0332],
+                  [-0.0137, -0.0332, -0.0205,  0.0000, -0.0205, -0.0332, -0.0137]]]])
 
     """
     KORNIA_CHECK_SHAPE(input, ["B", "C", "H", "W"])
diff --git a/kornia/filters/kernels.py b/kornia/filters/kernels.py
index de03927..bfeb4d8 100644
--- a/kornia/filters/kernels.py
+++ b/kornia/filters/kernels.py
@@ -74,6 +74,27 @@ def normalize_kernel2d(input: torch.Tensor) -> torch.Tensor:
     return input / (norm[..., None, None])
 
 
+def normalize_kernel2d_2nd_order(input: torch.Tensor) -> torch.Tensor:
+    r"""Scale a stack of second order derivative kernels ``(dxx, dxy, dyy)`` to derivative estimates.
+
+    Each kernel is divided by the magnitude of its response to the quadratic whose second derivative it
+    estimates, ``x**2 / 2`` for ``dxx``, ``x * y`` for ``dxy`` and ``y**2 / 2`` for ``dyy``, so the three
+    channels come out in the same units. Dividing every kernel by its own absolute sum instead, as
+    :func:`normalize_kernel2d` does, scales the mixed kernel differently from the pure ones.
+    """
+    KORNIA_CHECK_SHAPE(input, ["3", "H", "W"])
+
+    h, w = input.shape[-2:]
+    y = torch.arange(h, device=input.device, dtype=torch.float32) - (h - 1) / 2
+    x = torch.arange(w, device=input.device, dtype=torch.float32) - (w - 1) / 2
+    y, x = torch.meshgrid(y, x, indexing="ij")
+    quadratics = torch.stack([x * x / 2, x * y, y * y / 2])
+
+    norm = (input * quadratics).sum(dim=(-2, -1)).abs().to(input.dtype)
+
+    return input / norm[:, None, None]
+
+
 def gaussian(
     window_size: int,
     sigma: torch.Tensor | float,
diff --git a/kornia/filters/sobel.py b/kornia/filters/sobel.py
index 51cdb91..fa021c6 100644
--- a/kornia/filters/sobel.py
+++ b/kornia/filters/sobel.py
@@ -25,7 +25,12 @@ from torch import nn
 
 from kornia.core.check import KORNIA_CHECK_IS_TENSOR, KORNIA_CHECK_SHAPE
 
-from .kernels import get_spatial_gradient_kernel2d, get_spatial_gradient_kernel3d, normalize_kernel2d
+from .kernels import (
+    get_spatial_gradient_kernel2d,
+    get_spatial_gradient_kernel3d,
+    normalize_kernel2d,
+    normalize_kernel2d_2nd_order,
+)
 
 
 def spatial_gradient(input: torch.Tensor, mode: str = "sobel", order: int = 1, normalized: bool = True) -> torch.Tensor:
@@ -81,7 +86,7 @@ def spatial_gradient(input: torch.Tensor, mode: str = "sobel", order: int = 1, n
     else:
         kernel = get_spatial_gradient_kernel2d(mode, order, device=input.device, dtype=input.dtype)
         if normalized:
-            kernel = normalize_kernel2d(kernel)
+            kernel = normalize_kernel2d(kernel) if order == 1 else normalize_kernel2d_2nd_order(kernel)
 
     # prepare kernel
     b, c, h, w = input.shape
diff --git a/tests/feature/test_responces_local_features.py b/tests/feature/test_responces_local_features.py
index f0e87db..ceb9acf 100644
--- a/tests/feature/test_responces_local_features.py
+++ b/tests/feature/test_responces_local_features.py
@@ -303,22 +303,22 @@ class TestBlobHessian(BaseTester):
         expected = torch.tensor(
             [
                 [
-                    [-0.0564, -0.0759, -0.0342, -0.0759, -0.0564, -0.0057, 0.0000],
-                    [-0.0759, -0.0330, 0.0752, -0.0330, -0.0759, -0.0096, 0.0000],
+                    [-0.0137, -0.0332, -0.0342, -0.0332, -0.0137, -0.0010, 0.0000],
+                    [-0.0332, 0.0098, 0.0752, 0.0098, -0.0332, -0.0049, 0.0000],
                     [-0.0342, 0.0752, 0.1914, 0.0752, -0.0342, -0.0068, 0.0000],
-                    [-0.0759, -0.0330, 0.0752, -0.0330, -0.0759, -0.0096, 0.0000],
-                    [-0.0564, -0.0759, -0.0342, -0.0759, -0.0564, -0.0057, 0.0000],
-                    [-0.0057, -0.0096, -0.0068, -0.0096, -0.0057, -0.0005, 0.0000],
+                    [-0.0332, 0.0098, 0.0752, 0.0098, -0.0332, -0.0049, 0.0000],
+                    [-0.0137, -0.0332, -0.0342, -0.0332, -0.0137, -0.0010, 0.0000],
+                    [-0.0010, -0.0049, -0.0068, -0.0049, -0.0010, 0.0000, 0.0000],
                     [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
                 ],
                 [
-                    [-0.0564, -0.0522, -0.0522, -0.0564, -0.0057, 0.0000, 0.0000],
-                    [-0.0522, 0.0688, 0.0688, -0.0123, 0.0033, -0.0057, -0.0005],
-                    [-0.0522, 0.0688, -0.0755, -0.1111, -0.0123, -0.0564, -0.0057],
-                    [-0.0564, -0.0123, -0.1111, -0.0755, 0.0688, -0.0522, -0.0080],
-                    [-0.0057, 0.0033, -0.0123, 0.0688, 0.0688, -0.0522, -0.0080],
-                    [0.0000, -0.0057, -0.0564, -0.0522, -0.0522, -0.0564, -0.0057],
-                    [0.0000, -0.0005, -0.0057, -0.0080, -0.0080, -0.0057, -0.0005],
+                    [-0.0137, -0.0332, -0.0332, -0.0137, -0.0010, 0.0000, 0.0000],
+                    [-0.0332, 0.0820, 0.0693, -0.0039, 0.0117, -0.0010, 0.0000],
+                    [-0.0332, 0.0693, 0.0137, -0.0352, -0.0039, -0.0137, -0.0010],
+                    [-0.0137, -0.0039, -0.0352, 0.0137, 0.0693, -0.0332, -0.0059],
+                    [-0.0010, 0.0117, -0.0039, 0.0693, 0.0820, -0.0332, -0.0059],
+                    [0.0000, -0.0010, -0.0137, -0.0332, -0.0332, -0.0137, -0.0010],
+                    [0.0000, 0.0000, -0.0010, -0.0059, -0.0059, -0.0010, 0.0000],
                 ],
             ],
             device=device,
diff --git a/tests/feature/test_scale_space_detector.py b/tests/feature/test_scale_space_detector.py
index 9667a01..1ed59f0 100644
--- a/tests/feature/test_scale_space_detector.py
+++ b/tests/feature/test_scale_space_detector.py
@@ -256,8 +256,15 @@ class TestScaleSpaceDetector(BaseTester):
         if dtype in (torch.float16, torch.bfloat16):
             # The shifted response is flat at half precision and yields no maxima at all.
             pytest.skip("a Hessian response offset by 1.0 has no resolution left in half precision")
-        torch.manual_seed(0)
-        inp = torch.rand(1, 1, 64, 64, device=device, dtype=dtype)
+        # Gaussian blobs of increasing size give the Hessian one scale-space maximum per blob. Random noise
+        # left one or two maxima whose survival depended on the exact response scale and on the dtype.
+        coords = torch.arange(96, device=device, dtype=torch.float64)
+        yy, xx = torch.meshgrid(coords, coords, indexing="ij")
+        inp = torch.zeros(96, 96, device=device, dtype=torch.float64)
+        for i, sigma in enumerate((1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 5.0, 6.0, 7.0)):
+            cy, cx = 16 + 32 * (i // 3), 16 + 32 * (i % 3)
+            inp += torch.exp(-((yy - cy) ** 2 + (xx - cx) ** 2) / (2 * sigma**2))
+        inp = inp[None, None].to(dtype)
         det = ScaleSpaceDetector(50, resp_module=NegatedHessian()).to(device, dtype)
         lafs, resps = det(inp)
         filled = lafs[0].ne(0).any(dim=-1).any(dim=-1)
diff --git a/tests/filters/test_sobel.py b/tests/filters/test_sobel.py
index 788b410..ae96347 100644
--- a/tests/filters/test_sobel.py
+++ b/tests/filters/test_sobel.py
@@ -254,6 +254,27 @@ class TestSpatialGradient(BaseTester):
         edges = spatial_gradient(inp, "diff", normalized=True)
         self.assert_close(edges, expected)
 
+    @pytest.mark.parametrize("mode", ["sobel", "diff"])
+    def test_second_order_normalized_quadratics(self, mode, device, dtype):
+        # With normalized=True every second order channel estimates the derivative itself, so on a quadratic
+        # surface the channels are the exact second derivatives up to the sign convention of the kernels, and
+        # dxx * dyy - dxy**2 is the exact determinant of the Hessian. Before, dxy had a scale of its own.
+        coords = torch.arange(9, device=device, dtype=dtype)
+        y, x = torch.meshgrid(coords, coords, indexing="ij")
+        surfaces = torch.stack([x * x, x * y, y * y, x * x + y * y, (x + y) ** 2 / 2])[:, None]
+        out = spatial_gradient(surfaces, mode, order=2, normalized=True)[..., 3:-3, 3:-3]
+
+        expected_abs = torch.tensor(
+            [[2.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 2.0], [2.0, 0.0, 2.0], [1.0, 1.0, 1.0]],
+            device=device,
+            dtype=dtype,
+        )
+        self.assert_close(out.abs(), expected_abs[:, None, :, None, None].expand_as(out))
+
+        det = out[:, :, 0] * out[:, :, 2] - out[:, :, 1] ** 2
+        expected_det = torch.tensor([0.0, -1.0, 0.0, 4.0, 0.0], device=device, dtype=dtype)
+        self.assert_close(det, expected_det[:, None, None, None].expand_as(det))
+
     def test_noncontiguous(self, device, dtype):
         batch_size = 3
         inp = torch.rand(3, 5, 5, device=device, dtype=dtype).expand(batch_size, -1, -1, -1)
```

</details>

AI disclosure: I used an AI coding agent to help write this patch, the tests and this comment. I have read the change and the tests myself and I will answer review comments personally.

**Comment by ducha-aiki** [comment]
@Nicholas022400701 yes please, open it. There's no PR for this yet; since we can't assign outside contributors, this comment is the claim. The numbers in the issue are unchanged on `main` (`4d645c01c`) after #4683.

As the issue says, this changes `hessian_response` / `BlobHessian` values, so it needs a `changelog.d/<PR>.breaking.md` fragment that states the old scaling as well as the new one.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).
